src/Yesod/Page.hs

```diff
@@ -72,151 +62,79 @@ data PageConfig a position = PageConfig
-      Left err -> fail err
-      Right rawJson -> case eitherDecode $ BSL.fromStrict rawJson of
-        Left err -> fail err
-        Right value -> withObject "Cursor" parseCursor value
-   where
-    parseCursor o = Cursor
-      <$> o .: "path"
-      <*> o .: "params"
-      <*> o .: "position"
-      <*> (o .:? "limit")
-
-data Position position = First | Previous position | Next position | Last
+data Position position = First | Next position
 
 instance FromJSON p => FromJSON (Position p) where
   parseJSON = withObject "Position" $ \o -> do
     position <- o .: "position"
     case position :: Text of
       "first" -> pure First
-      "previous" -> Previous <$> o .: "keySet"
       "next" -> Next <$> o .: "keySet"
-      "last" -> pure Last
       unexpected -> fail $ show unexpected
 
 instance ToJSON p => ToJSON (Position p) where
   toJSON = \case
     First -> object ["position" .= ("first" :: Text)]
-    Previous p -> object ["position" .= ("previous" :: Text), "keySet" .= p]
     Next p -> object ["position" .= ("next" :: Text), "keySet" .= p]
-    Last -> object ["position" .= ("last" :: Text)]
 
-getPaginated
-  :: ( MonadHandler m
-     , ToJSON position
-     , FromJSON position
-     , ToJSON params
-     , FromJSON params
-     , RenderRoute (HandlerSite m)
-     )
-  => PageConfig a position
-  -> ParseParamM params
-  -> m (Cursor params position, [a] -> Page a)
-getPaginated pageConfig parser = do
-  route <- maybe (throwString "no route") pure =<< getCurrentRoute
-  cursor <- runParseParams pageConfig route parser
-  pure (cursor, withCursor pageConfig cursor)
-
-withCursor
-  :: (ToJSON params, ToJSON position)
-  => PageConfig a position
-  -> Cursor params position
-  -> [a]
-  -> Page a
-withCursor pageConfig cursor items = Page
-  { pageData = items
-  , pageFirst = makeCursor First
-  , pagePrevious = do
-    guard $ case cursorPosition cursor of
-      First -> False
-      Previous _ -> True
-      Next _ -> True
-      Last -> True
-    Just $ makeCursor . Previous $ toJSON mFirstId
-  , pageNext = do
-    guard . not $ null items || maybe False (len <) (cursorLimit cursor)
-    Just $ makeCursor . Next $ toJSON mLastId
-  , pageLast = makeCursor Last
-  }
- where
-  (len, mFirstId, mLastId) = unwrap $ foldMap wrap items
-  wrap x =
-    ( 1
-    , Monoid.First . Just $ makePosition pageConfig x
-    , Monoid.Last . Just $ makePosition pageConfig x
-    )
-  unwrap (s, f, l) = (getSum s, getFirst f, getLast l)
-  makeCursor position = Cursor
-    { cursorPath = cursorPath cursor
-    , cursorParams = toJSON $ cursorParams cursor
-    , cursorPosition = position
-    , cursorLimit = cursorLimit cursor
-    }
+cursorRouteAtPosition
+  :: ToJSON position => Cursor position -> Position position -> RenderedRoute
+cursorRouteAtPosition cursor = \case
+  First -> withNext Nothing
+  Next p -> withNext $ Just $ encodeText p
+  where withNext mNext = updateQueryParameter "next" mNext $ cursorRoute cursor
```

Comment:
Any thoughts on changing `"next"` to `"cursor"` or `"position"`? I think _next_ is a misnomer.

Reply:
Yeah, I have no dog in this fight -- you tell me :)

FWIW, my thought process was:

- `cursor`: wrong (IMO) since it's no longer `Cursor` being encoded
- `next`: made sense to me since it only ever appears as `next=p` of a `Next p`
- `position`: fine with me since `Next p :: Position position` so `p :: position`.

:man_shrugging: WDYT?

Reply:
I'm cool with `position`.